### PR TITLE
Adding support for VIP Removal

### PR DESCRIPTION
**Pull request by rccarper**
*Description of changes:*
Adding support for VIPs being removed when their TTL expires.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/s3_client.c [hunk 1]
```diff
@@ -110,20 +111,25 @@ static void s_s3_client_on_acquire_http_connection(
     void *user_data);
 
 /* Schedule task for processing work. */
-static void s_s3_client_schedule_process_work_task_synced(struct aws_s3_client *client);
+static void s_s3_client_schedule_process_work_synced(struct aws_s3_client *client);
+
```
Comment (justinboswell) [inline source/s3_client.c:115]: What is the difference between these two functions? One is commented, one is not, and says default in it :)

**Review by justinboswell — COMMENTED** [review]
If I'm reading this right, we no longer have a TTL override for these records. That will possibly limit the available VIPs to the point where it will affect throughput. Have you tested that?

File: source/s3_client.c (hunk 1 above)
Comment (rccarper) [reply source/s3_client.c:115]: Default is the actual function in the vtable, the other one calls the vtable.  I'll add a comment for that!

**Review by justinboswell — APPROVED** [review]

File: source/s3_client.c [hunk 2]
```diff
@@ -363,33 +383,31 @@ static void s_s3_client_start_destroy(void *user_data) {
...
-            aws_s3_vip_start_destroy(vip);
-            vip = NULL;
-        }
+        struct aws_s3_vip *vip = AWS_CONTAINER_OF(vip_node, struct aws_s3_vip, node);
 
-        s_s3_client_lock_synced_data(client);
-        s_s3_client_schedule_process_work_task_synced(client);
-        s_s3_client_unlock_synced_data(client);
+        aws_s3_vip_start_destroy(vip);
+        vip = NULL;
```
Comment (bretambrose) [inline source/s3_client.c:401]: Seems useless

File: source/s3_client.c [hunk 3]
```diff
@@ -1926,13 +1905,173 @@ static void s_s3_client_on_host_resolver_address_resolved(
...
+    for (size_t address_index = 0; address_index < aws_array_list_length(host_addresses); ++address_index) {
+
+        if (client->synced_data.active_vip_count >= client->ideal_vip_count) {
+            break;
+        }
+
+        if (client->synced_data.allocated_vip_count >= aws_s3_client_get_max_allocated_vip_count(client)) {
+            AWS_LOGF_WARN(
+                AWS_LS_S3_CLIENT,
+                "id=%p Allocated VIP count (%d) is greater than or equal to twice the amount of allowed VIPs (%d). "
```
Comment (bretambrose) [inline source/s3_client.c:1960]: text here (twice) is kind of couple to the underlying, current threshold, could be reworded to be less likely to be come inaccurate in the future.

File: source/s3_client.c [hunk 4]
```diff
@@ -1926,13 +1905,173 @@ static void s_s3_client_on_host_resolver_address_resolved(
...
+        aws_linked_list_push_back(&client->synced_data.vips, &vip->node);
+
+        AWS_LOGF_INFO(
+            AWS_LS_S3_CLIENT,
+            "id=%p Initiating creation of VIP with address '%s', total active vip count %d",
+            (void *)client,
+            (const char *)host_address_byte_cursor.ptr,
+            client->synced_data.active_vip_count);
+
+        s_s3_client_schedule_process_work_synced(client);
```
Comment (bretambrose) [inline source/s3_client.c:2016]: This being called each iteration is a little odd.

**Review by bretambrose — APPROVED** [review]

**Pull request merged**